diff --git original/drivers/media/platform/mxc/capture/ov5645_mipi_v2.c modified/drivers/media/platform/mxc/capture/ov5645_mipi_v2.c
index 0052e97..4346cb6 100644
--- original/drivers/media/platform/mxc/capture/ov5645_mipi_v2.c
+++ modified/drivers/media/platform/mxc/capture/ov5645_mipi_v2.c
@@ -3315,11 +3315,11 @@ static int ov5645_s_parm(struct v4l2_subdev *sd, struct v4l2_streamparm *a)
 		   It is faster to skip calling it, however this is buggy.
 		   More work needed here. Define OV5645_MIPI_FASTER_INIT
 		   in order to skip sensor init unless the mode changes */
-		if ((orig_mode != (u32)a->parm.capture.capturemode) ||
+		if ((orig_mode != sensor->streamcap.capturemode) ||
 				(frame_rate != sensor->loaded_fps)) {
 #endif // OV5645_MIPI_FASTER_INIT
 			ret = ov5645_init_mode(sensor, frame_rate,
-					(u32)a->parm.capture.capturemode, orig_mode);
+					sensor->streamcap.capturemode, orig_mode);
 #ifdef OV5645_MIPI_FASTER_INIT
 		}
 #endif // OV5645_MIPI_FASTER_INIT
@@ -3327,9 +3327,7 @@ static int ov5645_s_parm(struct v4l2_subdev *sd, struct v4l2_streamparm *a)
 			return ret;
 
 		sensor->streamcap.timeperframe = *timeperframe;
-		sensor->streamcap.capturemode =
-				(u32)a->parm.capture.capturemode;
-		sensor->loaded_mode = a->parm.capture.capturemode;
+		sensor->loaded_mode = sensor->streamcap.capturemode;
 		sensor->loaded_fps = frame_rate;
 		break;
 
@@ -3541,11 +3539,34 @@ static int ov5645_s_stream(struct v4l2_subdev *sd, int enable)
 {
 	struct i2c_client *client = v4l2_get_subdevdata(sd);
 	struct ov5645 *sensor = to_ov5645(client);
+	enum ov5645_frame_rate frame_rate;
+	u32 tgt_fps;
+	int ret;
 
-	if (enable)
-		OV5645_stream_on(sensor);
-	else
+	if (!enable) {
 		OV5645_stream_off(sensor);
+		return 0;
+	}
+
+	tgt_fps = sensor->streamcap.timeperframe.denominator /
+		  sensor->streamcap.timeperframe.numerator;
+	if (tgt_fps == 15)
+		frame_rate = ov5645_15_fps;
+	else if (tgt_fps == 30)
+		frame_rate = ov5645_30_fps;
+	else if (tgt_fps == 60)
+		frame_rate = ov5645_60_fps;
+	else
+		return -EINVAL;
+
+	ret = ov5645_init_mode(sensor, frame_rate,
+		sensor->streamcap.capturemode, sensor->loaded_mode);
+	if (ret < 0)
+		return ret;
+
+	sensor->loaded_mode = sensor->streamcap.capturemode;
+	sensor->loaded_fps = frame_rate;
+	OV5645_stream_on(sensor);
 
 	return 0;
 }
